File: Homework/themoviedb/src/main/java/com/example/themoviedb/recyclerView/SearchMovieAdapter.java
package com.example.themoviedb.recyclerView;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.themoviedb.Injection;
import com.example.themoviedb.MovieDataSource;
import com.example.themoviedb.R;
import com.example.themoviedb.Utilities;
import com.example.themoviedb.models.MovieModel;
import com.example.themoviedb.persistence.Movie;
import com.google.android.material.snackbar.Snackbar;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

import io.reactivex.android.schedulers.AndroidSchedulers;

public class SearchMovieAdapter extends RecyclerView.Adapter<SearchMovieAdapter.SearchMovieViewHolder> {
    private ArrayList<MovieModel> mMovies;

    public SearchMovieAdapter(){
        mMovies = new ArrayList<>();
    }

    public void setMovies(ArrayList<MovieModel> movies){
        mMovies.clear();
        mMovies.addAll(movies);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public SearchMovieViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.search_movie_card, viewGroup, false);
        return new SearchMovieViewHolder(view);
    }

    @SuppressLint("CheckResult")
    @Override
    public void onBindViewHolder(@NonNull SearchMovieViewHolder viewHolder, int i) {
        final MovieModel currentItem = mMovies.get(i);

        Picasso.get()
                .load(Utilities.getMoviePosterLink(viewHolder.context, currentItem.posterPath))
                .into(viewHolder.moviePoster);
        viewHolder.titleTextView.setText(currentItem.title);

        MovieDataSource dataSource = Injection.provideMovieDataSource(viewHolder.context);

        dataSource.getMovie(currentItem.title)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(addedMovie -> viewHolder.addRemoveButton.setImageResource(addedMovie != null
                        ? R.drawable.check
                        : R.drawable.plus));

        viewHolder.addRemoveButton.setOnClickListener((view)->{
            final ImageButton castedView = (ImageButton)view;

            dataSource.getMovie(currentItem.title)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(persistMovie -> {
                        int snackBarText;
                        if (persistMovie != null){
                            castedView.setImageResource(R.drawable.plus);
                            dataSource.deleteMovie(persistMovie);
                            snackBarText = R.string.movieRemoved;
                        } else{
                            castedView.setImageResource(R.drawable.check);
                            Movie newMovie = new Movie(currentItem);
                            dataSource.insertOrUpdateMovie(newMovie);
                            snackBarText = R.string.movieAdded;
                        }

                        Snackbar snackbar = Snackbar.make(view, snackBarText, Snackbar.LENGTH_LONG);
                        snackbar.setAction(R.string.undoString, new UndoListener());
                        snackbar.show();
                    });
        });
    }

    public class UndoListener implements View.OnClickListener {

        @Override
        public void onClick(View v) {
            // Code to undo the user's last action
        }
    }

    @Override
    public int getItemCount() {
        return mMovies.size();
    }

    static class SearchMovieViewHolder extends RecyclerView.ViewHolder {
        TextView titleTextView;
        ImageView moviePoster;
        ImageButton addRemoveButton;
        Context context;

        SearchMovieViewHolder(View itemView) {
            super(itemView);
            titleTextView = itemView.findViewById(R.id.movieTitle);
            moviePoster = itemView.findViewById(R.id.moviePoster);
            addRemoveButton = itemView.findViewById(R.id.addRemoveButton);
            context = itemView.getContext();
        }
    }
}

File: Homework/themoviedb/src/main/java/com/example/themoviedb/recyclerView/SearchMovieAdapter.kt
package com.example.themoviedb.recyclerView

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.themoviedb.Injection
import com.example.themoviedb.R
import com.example.themoviedb.Utilities
import com.example.themoviedb.models.MovieModel
import com.example.themoviedb.persistence.Movie
import com.google.android.material.snackbar.Snackbar
import io.reactivex.CompletableObserver
import io.reactivex.disposables.Disposable
import io.reactivex.observers.DisposableMaybeObserver
import io.reactivex.schedulers.Schedulers
import java.util.*



class SearchMovieAdapter : RecyclerView.Adapter<SearchMovieAdapter.SearchMovieViewHolder>() {
    private val mMovies: ArrayList<MovieModel> = ArrayList()

    fun setMovies(movies: ArrayList<MovieModel>?) {
        mMovies.clear()
        mMovies.addAll(movies!!)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): SearchMovieViewHolder {
        val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.search_movie_card, viewGroup, false)
        return SearchMovieViewHolder(view)
    }

    @SuppressLint("CheckResult")
    override fun onBindViewHolder(viewHolder: SearchMovieViewHolder, i: Int) {
        val currentItem = mMovies[i]

        Utilities.setImageResource(viewHolder.context, viewHolder.moviePoster, currentItem.posterPath)
        viewHolder.titleTextView.text = currentItem.title

        val dataSource = Injection.provideMovieDataSource(viewHolder.context)

        dataSource.getMovie(currentItem.title)
                .subscribeOn(Schedulers.io())
                .subscribe { addedMovie ->
                    viewHolder.addRemoveButton.setImageResource(if (addedMovie != null)
                        R.drawable.check
                    else
                        R.drawable.plus)
                }

        viewHolder.addRemoveButton.setOnClickListener { view ->
            val castedView = view as ImageButton

            dataSource.getMovie(currentItem.title)
                    .subscribeOn(Schedulers.io())
                    .subscribe(object : DisposableMaybeObserver<Movie>() {
                        override fun onSuccess(movie: Movie) {
                            dataSource.deleteMovie(movie)
                                    .subscribeOn(Schedulers.io())
                                    .subscribe(object : CompletableObserver {
                                        override fun onError(e: Throwable) {
                                        }
                                        override fun onSubscribe(d: Disposable) {
                                        }
                                        override fun onComplete() {
                                            castedView.setImageResource(R.drawable.plus)
                                            showSnackBar(view, R.string.movieRemoved, View.OnClickListener {
                                                val newMovie = Movie(currentItem)
                                                dataSource.insertMovie(newMovie)
                                                        .subscribeOn(Schedulers.io())
                                                        .subscribe(object : CompletableObserver{
                                                            override fun onError(e: Throwable) {
                                                            }
                                                            override fun onSubscribe(d: Disposable) {
                                                            }
                                                            override fun onComplete() {
                                                                castedView.setImageResource(R.drawable.check)
                                                            }
                                                        })
                                            })
                                        }
                                    })
                        }

                        override fun onComplete() {
                            val newMovie = Movie(currentItem)
                            dataSource.insertMovie(newMovie)
                                    .subscribeOn(Schedulers.io())
                                    .subscribe(object : CompletableObserver{
                                        override fun onError(e: Throwable) {
                                        }
                                        override fun onSubscribe(d: Disposable) {
                                        }
                                        override fun onComplete() {
                                            castedView.setImageResource(R.drawable.check)
                                            showSnackBar(view, R.string.movieAdded, View.OnClickListener {
                                                dataSource.getMovie(currentItem.title)
                                                        .subscribeOn(Schedulers.io())
                                                        .subscribe(object : DisposableMaybeObserver<Movie>() {
                                                            override fun onSuccess(addedMovie: Movie) {
                                                                dataSource.deleteMovie(addedMovie)
                                                                        .subscribeOn(Schedulers.io())
                                                                        .subscribe(object : CompletableObserver {
                                                                            override fun onError(e: Throwable) {
                                                                            }
                                                                            override fun onSubscribe(d: Disposable) {
                                                                            }
                                                                            override fun onComplete() {
                                                                                castedView.setImageResource(R.drawable.plus)
                                                                            }
                                                                        })
                                                            }
                                                            override fun onComplete() {
                                                            }
                                                            override fun onError(e: Throwable) {
                                                            }
                                                        })
                                            })
                                        }
                                    })
                        }

                        override fun onError(e: Throwable) {
                        }
                    })
        }
    }

    fun showSnackBar(view: View, titleResourceId: Int, onClickListener: View.OnClickListener){
        val snackBar = Snackbar.make(view, titleResourceId, Snackbar.LENGTH_LONG)
        snackBar.setAction(R.string.undoString, onClickListener)
        snackBar.show()
    }

    override fun getItemCount(): Int {
        return mMovies.size
    }

    class SearchMovieViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var titleTextView: TextView = itemView.findViewById(R.id.movieTitle)
        var moviePoster: ImageView = itemView.findViewById(R.id.moviePoster)
        var addRemoveButton: ImageButton = itemView.findViewById(R.id.addRemoveButton)
        var context: Context = itemView.context
    }
}